### ISSUE TITLE
Default shader doesn't work on all hardware

### ISSUE THREAD

**Issue by darrylryan** [Zylann/godot_heightmap_native_plugin#22]
The terrain is black on my laptop... after playing with the shader and stripping it back to the basic albedo splatmap it works, so I think there must be something not quite right in the shader.

**Comment by Zylann** [comment]
The current shader is unoptimized, it samples way to many textures, so I expect it to run on mostly high-end hardware for now. I plan to reduce the amount of texture fetches in later updates.

Did you see any error? Were you able to identify exactly which part makes it fail?

**Comment by darrylryan** [comment]
Yes I think it could be considered an engine bug since there are no errors with the shaders, it just fails to work.
It's a Radeon HD7670M I'm using.

**Comment by Zylann** [comment]
Can you identify at which point the shader fails when you modify it? I can't reproduce this on my computer

**Comment by Zylann** [comment]
Yeah, that's what I thought... some old graphics cards may not support many textures.
I plan to reduce them: I already removed the separate hold map to be in the color map, and I should merge roughness with albedo and bump with normal map. That means extra texture baking to be done, though.

**Comment by Zylann** [comment]
Can you try again? I submitted a change that makes detail textures packed, so now there should be 11 texture fetches in the fragment shader, instead of 19.

**Comment by anubhabg** [comment]
Unfortunately it still does not work with same error.
The maximum number of textures should be defined by gl.MAX_TEXTURE_IMAGE_UNITS according to this [StackOverflow answer](https://stackoverflow.com/a/20850584) which on my PC is 16 as reported by glxinfo (Is this limit same in GLES too?).

Godot dumped the whole generated 1735 line GLSL code in console [pastebin](https://pastebin.com/gWakHJGj). Your fragment code starts at around line 1189. (I hardcoded depth_blending to false at line 1211.)

So I think the problem is because Godot uses more texture samplers internally (described at line 21). I think perhaps an easier solution would be to provide a different simple shader without roughness, depth_blending etc. as most WebGL (mobile?) platforms seem to support only 16 ([stats from here](https://webglstats.com/webgl2/parameter/MAX_TEXTURE_IMAGE_UNITS)).

**Comment by Zylann** [comment]
Wow, that graphics card you have sounds quite limitating :(
Also note that now, there is only two textures per slot in the terrain, so removing roughness, bump or depth blending won't change anything because those are actually part of the ALBEDO and NORMAL textures (it's a single texture fetch for both).

Ooh hang on... in the pastebin you have me, the shader still has separate roughness Oo
Those shouldn't be here, they are not used.

**Comment by darrylryan** [comment]
Yep, sorry for the slow reply, but I also have the same situation - commenting out the extra textures makes it work.

I think we really need those texture arrays in the core...  to complement this terrain we could do with tree billboards and grass for a foliage system. That would be much easier to do if all the different trees could each have their own texture in the array, and each texture could contain 8 or so different view angles of the tree.

**Comment by Zylann** [comment]
@darrylryan why trees? Trees aren't part of the terrain material, they can work fine without texture arrays.

**Comment by anubhabg** [comment]
It's working now out of the box with depth blending! I guess yesterday the shader being open in Godot wasn't updated by git. Performance is poor though (at least of the demo) but that should be expected from really old hardware like this.
Sadly it appears a lot of Android devices have MAX_TEXTURE_IMAGE_UNITS = 8 from [here](http://opengles.gpuinfo.org/displaycapability.php?name=GL_MAX_TEXTURE_IMAGE_UNITS&esversion=2).

**Comment by darrylryan** [comment]
@Zylann I might have a go at making some basic tree imposters for this terrain system...  the two things need to work together because the trees and grass need to be placed on the terrain, so need to know how to read the height and ideally load the grass only for active chunks of terrain.

Do you plan to support paging/streaming large terrains with this?

**Comment by Zylann** [comment]
@darrylryan you can get height with `get_interpolated_height_at()`. For trees I was thinking of something a bit wider, like any mesh (rocks too). Grass is too numerous to use the same technique. For streaming, I have notes but it's a massive change to do in the internal logic, I need to prototype stuff. The thing is, the API limitation of Godot makes it quite "prehistoric" compared to what games like FC5 do xD but those things are different issues to be discussed in another thread.

**Comment by Zylann** [comment]
I managed to spare 2 more texture fetches in the fragment shader. It's not much but maybe that will improve a little.
There is a way to spare another 2 but I'll need more time to figure out how to do it properly (the amount of varyings may increase too, which is also a limit according to WebGL stats :| )

**Comment by Zylann** [comment]
Some news with commit 19ac52868d751de2b45c6b4cf12d37a64032d563:

- There is now a lite version of the shader, at the moment it has to be toggled from within hterrain.gd, near the beginning of the script. I'll have to work on easier shader swapping in the future
- Bump and roughness have been swapped, so the lite shader can perform depth blending without too much plugin modifications 
- Due to the above, the demo textures won't work anymore and will need to be re-packed and re-assigned (use an image editor, or the channel packer plugin I made. Or just leave them white if you don't need depth blending)(

**Comment by Zylann** [comment]
Current master build now has easy dropdown to select a less-demanding shader (see also previous comment)
